src/user_account/models.py

```diff
@@ -23,10 +25,18 @@ def total_score(self):
         return self.avr_score
 
     def last_run(self):
-        return self.test_results.last().datetime_run
+        if self.test_results.count() != 0:
+            return self.test_results.last().datetime_run
+        else:
+            return "_____"
 
     def percent_success_passed(self):
-        percent_success_passed = sum([user.percent_correct_answers() for user in
-                                      self.test_results.filter(is_completed=True)])
-
-        return round(percent_success_passed, 2)
+        count_passed_test = self.count_passed_tests()
+        test_results = self.test_results.filter(is_completed=True)
+
+        if count_passed_test != 0:
+            percent_success_passed = round((sum([test_result.correct_answers_count() for test_result in test_results]) / \
+                                            sum([test_result.test_question_count() for test_result in test_results]))
+                                           * 100, 2)
+            return percent_success_passed
```

Comment:
Technically, it is correct, but leads to long-running leaderboard page rendering because of heavy sql-queries.

This is where signals could also help. 

For this, `percent_success_passed` attribute could be added in a user model. Then, via signal subscription the attribute is updated when a test run is finished. In this case, the attribute will act as a cache for heavily calculated value. 

As a result, reading attribute as frequent operation will be cheap, but writing to attribute as an expensive operation will happen only when it is absolutely necessary.

Reply:
Ready. Thanks a lot!